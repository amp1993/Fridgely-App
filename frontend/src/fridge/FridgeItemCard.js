import React, { useState, useEffect } from "react";
import { useNavigate } from "react-router-dom";
import FridgelyApi from "../api/api";
import "./FridgeItemCard.css"


// FridgeItemCard component to display, increase, decrease, and delete items in the fridge

const FridgeItemCard = ({ fridgeItems, user, updateFridgeItems }) => {
  const [sortedItems, setSortedItems] = useState([]);
  const navigate = useNavigate()

  useEffect(() => {
    // Flatten the nested array of items
    const flattenedItems = fridgeItems[0];
    // Sort the flattened array by categoryName
    const sorted = flattenedItems.sort((a, b) => a.categoryName.localeCompare(b.categoryName));

    // Set the sorted items in state
    setSortedItems(sorted);
  }, [fridgeItems]);

  // Group the sorted items by categoryName
  const groupedItems = sortedItems.reduce((acc, item) => {
    const category = item.categoryName;
    if (!acc[category]) {
      acc[category] = [];
    }
    acc[category].push(item);
    return acc;
  }, {});



  // Handle quantity increase
  const handleIncrease = async (productName) => {

    const currentItem = fridgeItems[0].find(item => item.productName === productName);
    if (!currentItem) {
      console.error("Item not found");
      return;
    }

    // Increment the quantity
    const newQuantity = parseInt(currentItem.quantityInFridge, 10) + 1;

    try {
      // Update the item in the backend
      await FridgelyApi.updateQtyFridgeItem({ quantityInFridge: newQuantity }, productName, user);

      // Update the state with the updated item
      const fetchUpdatedData = await FridgelyApi.getFridgeItems(user);
      updateFridgeItems(fetchUpdatedData);

    } catch (error) {
      console.error("Error updating item:", error);
    }
  };

  // Handle quantity decrease
  const handleDecrease = async (productName) => {

    const currentItem = fridgeItems[0].find(item => item.productName === productName);
    if (!currentItem) {
      console.error("Item not found");
      return;
    }
    // Decrement the quantity but ensure it doesn't go below 0

    const newQuantity = parseInt(Math.max(currentItem.quantityInFridge - 1, 0));
    try {
      await FridgelyApi.updateQtyFridgeItem({ quantityInFridge: newQuantity }, productName, user);
      const fetchUpdatedData = await FridgelyApi.getFridgeItems(user);
      updateFridgeItems(fetchUpdatedData);
    } catch (error) {
      console.error("Error updating item:", error);
    }
  };

  // Handle item deletion
  const handleDelete = async (productName) => {
    try {
      await FridgelyApi.removeItemFromFridge(productName, user);
      const fetchUpdatedData = await FridgelyApi.getFridgeItems(user);
      updateFridgeItems(fetchUpdatedData);

    } catch (error) {
      console.error("Error deleting item:", error);
    }
  };
  // Navigate to new item form
  const addItem = () => {
    navigate(`/fridge/${user}`)
  }
  return (
    <div className="item-list-container">
      <button className="add-item-button" onClick={() => addItem()}>Add an Item</button>

      {Object.keys(groupedItems).map(category => (
        <div key={category}>
          <h3>{category}</h3>
          <ul className="item-list">
            {groupedItems[category].map(fridgeItem => (
              <li className="item-item" key={fridgeItem.id}>

                <p className="item-text"><strong>{fridgeItem.productName}</strong></p>
                <div className="item-buttons">
                  <button onClick={() => handleDecrease(fridgeItem.productName)}>←</button>
                  <p style={{ margin: '0 10px' }}>
                    {fridgeItem.quantityInFridge} {fridgeItem.unitOfMeasure}
                  </p>
                  <button onClick={() => handleIncrease(fridgeItem.productName)}>→</button>
                  <button className="delete-button" onClick={() => handleDelete(fridgeItem.productName)}>Delete</button>
                </div>
              </li>
            ))}
          </ul>
        </div>
      ))}

    </div>
  );
};

export default FridgeItemCard;
