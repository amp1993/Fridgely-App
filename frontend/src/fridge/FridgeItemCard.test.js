import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import FridgeItemCard from './FridgeItemCard';

const mockFridgeItems = [
  { id: 1, productName: 'Apple', categoryName: 'Fruits', quantityInFridge: 5, unitOfMeasure: 'pcs' },
  { id: 2, productName: 'Carrot', categoryName: 'Vegetables', quantityInFridge: 10, unitOfMeasure: 'pcs' },
];

it('renders without crashing', function() {
  render(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );
});

it('matches snapshot', function() {
  const { asFragment } = render(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it('displays fridge items correctly', function() {
  const { getByText } = render(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  expect(getByText('Apple')).toBeInTheDocument();
  expect(getByText('Carrot')).toBeInTheDocument();
});

it('handles quantity increase and decrease correctly', async function() {
  const { getAllByText, getByText, rerender } = render(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  const increaseButtons = getAllByText('→');
  const decreaseButtons = getAllByText('←');
  const appleQuantity = getByText('5 pcs');

  fireEvent.click(increaseButtons[0]);

  // Mock the state update to immediately reflect the change
  mockFridgeItems[0].quantityInFridge += 1;
  rerender(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  await waitFor(() => expect(appleQuantity.textContent).toBe('6 pcs'));

  fireEvent.click(decreaseButtons[0]);

  // Mock the state update to immediately reflect the change
  mockFridgeItems[0].quantityInFridge -= 1;
  rerender(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  await waitFor(() => expect(appleQuantity.textContent).toBe('5 pcs'));
});

it('handles item deletion correctly', async function() {
  const { getAllByText, queryByText, rerender } = render(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[mockFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  const deleteButtons = getAllByText('Delete');
  fireEvent.click(deleteButtons[0]);

  // Mock the state update to immediately reflect the change
  const updatedFridgeItems = mockFridgeItems.filter(item => item.productName !== 'Apple');
  rerender(
    <BrowserRouter>
      <FridgeItemCard fridgeItems={[updatedFridgeItems]} user="testuser" updateFridgeItems={() => {}} />
    </BrowserRouter>
  );

  await waitFor(() => expect(queryByText('Apple')).not.toBeInTheDocument());
});
