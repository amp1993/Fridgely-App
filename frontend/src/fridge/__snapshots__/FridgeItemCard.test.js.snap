// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="item-list-container"
  >
    <button
      class="add-item-button"
    >
      Add an Item
    </button>
    <div>
      <h3>
        Fruits
      </h3>
      <ul
        class="item-list"
      >
        <li
          class="item-item"
        >
          <p
            class="item-text"
          >
            <strong>
              Apple
            </strong>
          </p>
          <div
            class="item-buttons"
          >
            <button>
              ←
            </button>
            <p
              style="margin: 0px 10px;"
            >
              5 pcs
            </p>
            <button>
              →
            </button>
            <button
              class="delete-button"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
    <div>
      <h3>
        Vegetables
      </h3>
      <ul
        class="item-list"
      >
        <li
          class="item-item"
        >
          <p
            class="item-text"
          >
            <strong>
              Carrot
            </strong>
          </p>
          <div
            class="item-buttons"
          >
            <button>
              ←
            </button>
            <p
              style="margin: 0px 10px;"
            >
              10 pcs
            </p>
            <button>
              →
            </button>
            <button
              class="delete-button"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
